cpp/include/cuspatial/cubic_spline.hpp

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <memory>
+#include <cudf/column/column.hpp>
+#include <cudf/table/table.hpp>
+
+namespace cuspatial {
+namespace detail {
+
+/**
+ * @brief Create a table of cubic spline coefficients from columns of coordinates.
+ *
+ * Computes coefficients for a natural cubic spline similar to the method
+ * found on http://mathworld.wolfram.com/CubicSpline.html .
+ *
+ * The input data arrays `t` and `y` contain the vertices of many concatenated
+ * splines.
+ *
+ * Currently, all input splines must be the same length. The minimum supported
+ * length is 5.
+ * 
+ * @note Ids should be prefixed with a 0, even when only a single spline
+ * is fit, ids will be {0, 0}
+ *
+ * @param[in] t column_view of independent coordinates for fitting splines
+ * @param[in] y column_view of dependent variables to be fit along t axis
+ * @param[in] ids of incoming coordinate sets
+ * @param[in] offsets the exclusive scan of the spline sizes, prefixed by
+ * 0. For example, for 3 splines of 5 vertices each, the offsets input array
+ * is {0, 5, 10, 15}.
+ * @param[in] mr the optional caller specified RMM memory resource
+ * @param[in] stream the optional caller specified cudaStream
+ *
+ * @return cudf::table_view of coefficients for spline interpolation. The size
+ * of the table is ((M-n), 4) where M is `t.size()` and and n is 
+ * `ids.size()-1`.
+**/
+std::unique_ptr<cudf::experimental::table> cubicspline_coefficients(
+                                         cudf::column_view const& t,
+                                         cudf::column_view const& y,
+                                         cudf::column_view const& ids,
+                                         cudf::column_view const& offsets,
+                                         rmm::mr::device_memory_resource *mr,
```

Comment:
In the non-detail API, `mr` should be defaulted to the default resource. Stream should not exist in the non-detail API, and be defaulted to zero in the detail API.

Reply:
They're just duped below.